lib/SILOptimizer/Transforms/SILLowerAggregateInstrs.cpp

```diff
@@ -109,17 +112,23 @@ static bool expandCopyAddr(CopyAddrInst *CA) {
     //   retain_value %new : $*T
     IsTake_t IsTake = CA->isTakeOfSrc();
     if (IsTake_t::IsNotTake == IsTake) {
-      TL.emitLoweredCopyValue(Builder, CA->getLoc(), New,
-                              TypeLowering::LoweringStyle::Deep);
+      if (expand)
+        TL.emitLoweredCopyValue(Builder, CA->getLoc(), New,
+                                TypeLowering::LoweringStyle::Deep);
+      else
+        TL.emitCopyValue(Builder, CA->getLoc(), New);
```

Comment:
I think you can clean this up if you fix a bug in TypeLowering.h. In TypeLowering.cpp. The specific problem is that emitLoweredCopyValue ignores the LoweringStyle flag and always performs a deep copy. If you fix that so that it just checks the lowering and does a copy_value if it is shallow, you could clean up the multiple instances of this code by creating a helper function that goes from expand to LoweringStyle:

```
static TypeLowering::LoweringStyle getLoweringStyle(bool shouldExpand) {
    if (shouldExpand)
       return TypeLowering::LoweringStyle::Deep;
    return TypeLowering::LoweringStyle::Shallow;
}
```

Then you could just write this code and hide the conditional:

```
TL.emitLoweredCopyValue(Builder, CA->getLoc(), New, getLoweringStyle(expand));
```

IMO, the conditional does not really add anything to the flow of the code and is just noise.

Reply:
Done! looks much better